modules/boulder/src/main/ts/ephox/boulder/core/StructureProcessor.ts

```diff
@@ -18,94 +16,104 @@ export interface StructureProcessor {
-          return SimpleResult.bind(
-            strictAccess(path, obj, key),
-            bundle
-          );
-        case FieldPresenceTag.DefaultedThunk:
-          return SimpleResult.bind(
-            fallbackAccess(obj, key, presence.process),
-            bundle
-          );
-        case FieldPresenceTag.Option:
-          return SimpleResult.bind(
-            optionAccess(obj, key),
-            bundleAsOption
-          );
-        case FieldPresenceTag.DefaultedOptionThunk:
-          return SimpleResult.bind(
-            optionDefaultedAccess(obj, key, presence.process),
-            bundleAsOption
-          );
-        case FieldPresenceTag.MergeWithThunk: {
-          const base = presence.process(obj);
-          const result = SimpleResult.map(
-            fallbackAccess(obj, key, Fun.constant({})),
-            (v) => Merger.deepMerge(base, v)
-          );
-          return SimpleResult.bind(result, bundle);
-        }
-      }
-    },
-    (newKey, instantiator) => {
-      const state = instantiator(obj);
-      return SimpleResult.svalue(ObjWriter.wrap(newKey, state));
+const extractField = <T, U>(
+  field: FieldPresence,
+  path: string[],
+  obj: Record<string, T>,
+  key: string,
+  prop: StructureProcessor
+): SimpleResult<SchemaError[], U | Optional<U>> => {
+  const bundle = (av: T): SimpleBundle<U> => prop.extract(path.concat([ key ]), av);
+
+  const bundleAsOption = (optValue: Optional<T>): OptionBundle<U> => optValue.fold(
+    () => SimpleResult.svalue(Optional.none()),
+    (ov) => {
+      const result = prop.extract(path.concat([ key ]), ov);
+      return SimpleResult.map(result, Optional.some);
     }
   );
+
+  switch (field.tag) {
+    case FieldPresenceTag.Required:
+      return requiredAccess(path, obj, key, bundle);
+    case FieldPresenceTag.DefaultedThunk:
+      return fallbackAccess(obj, key, field.process, bundle);
+    case FieldPresenceTag.Option:
+      return optionAccess(obj, key, bundleAsOption);
+    case FieldPresenceTag.DefaultedOptionThunk:
+      return optionDefaultedAccess(obj, key, field.process, bundleAsOption);
+    case FieldPresenceTag.MergeWithThunk: {
+      return fallbackAccess(obj, key, Fun.constant({}), (v) => {
+        const result = Merger.deepMerge(field.process(obj), v);
+        return bundle(result);
+      });
+    }
+  }
 };
 
-const cExtract = <T>(path: string[], obj: Record<string, T>, fields: FieldProcessor[]): SimpleResult<SchemaError[], T> => {
-  const results = Arr.map(fields, (field) => cExtractOne(path, obj, field));
-  return ResultCombine.consolidateObj(results, {});
+const extractFields = <T, U>(
+  path: string[],
+  obj: Record<string, T>,
+  fields: FieldProcessor[]
+): SimpleResult<SchemaError[], Record<string, U | Optional<U>>> => {
+  const success: Record<string, U | Optional<U>> = {};
+  const errors: SchemaError[] = [];
+
+  // PERFORMANCE: We use a for loop here instead of Arr.each as this is a hot code path
```

Comment:
Just because I'm curious: what sort of numbers did we get changing from a for loop to Arr.each here?

Reply:
As mentioned on slack, it was about ~1/6th of the savings provided in refactoring this part of boulder.